### PR TITLE
GH-36155: [C++][Go][Java][FlightRPC] Add support for long-running queries

### LINKED ISSUE #36155: [C++][Go][Java][FlightRPC] Add support for long-running queries
### Describe the enhancement requested

Based on the proposal in https://docs.google.com/document/d/1jhPyPZSOo2iy0LqIJVUs9KWPyFULVFJXTILDfkadx2g/edit# .
See also the discussion thread: https://lists.apache.org/thread/247z3t06mf132nocngc1jkp3oqglz7jp

In Flight RPC, FlightInfo includes addresses of workers alongside result partition info. This lets clients fetch data directly from workers[^1], in parallel or even distributed across multiple machines. But this also comes with tradeoffs.

[^1]: Of course, servers are free to return the location of a proxy/load balancer/etc., or omit locations and have the client fetch results from the same server that they issued the query to. Flight RPC offers this flexibility to servers; clients don’t have to know or care. 

Queries generally don’t complete instantly (as much as we would like them to). So where can we put the ‘query evaluation time’?

  * In GetFlightInfo: block and wait for the query to complete.
    * Con: this is a long-running blocking call, which may fail or time out. Then when the client retries, the server has to redo all the work.
    * Con: parts of the result may be ready before others, but the client can’t do anything until everything is ready.
  * In DoGet: return a fixed number of partitions
    * Con: this makes handling worker failures hard. Systems like Trino support [fault-tolerant execution](https://trino.io/docs/current/admin/fault-tolerant-execution.html) by replacing workers at runtime. But GetFlightInfo has already passed, so we can’t notify the client of new workers[^2].
    * Con: we have to know or fix the partitioning up front.

[^2]: Again, the server could proxy workers, or depend on Kubernetes DNS routing, or configure gRPC XDS. But this somewhat defeats the point of returning worker locations in the first place, and is much more complicated (operationally, implementation-wise).

Neither solution is optimal.

### Proposal

We can address this by adding a retryable version of GetFlightInfo. First, we add a new RPC call and result message:

```protobuf
service FlightService {
  // ...
  rpc PollFlightInfo(FlightDescriptor) returns (PollInfo);
}

message PollInfo {
  // The currently available results so far.
  FlightInfo info = 1;
  // The descriptor the client should use on the next try.
  // If unset, the query is complete.
  FlightDescriptor flight_descriptor = 2;
  // Query progress. Must be in [0.0, 1.0] but need not be
  // monotonic or nondecreasing. If unknown, do not set.
  optional double progress = 3;
  // Expiration time for this request. After this passes, the server
  // might not accept the retry descriptor anymore (and the query may 
  // be cancelled). This may be updated on a call to PollFlightInfo.
  google.protobuf.Timestamp expiration_time = 4;
}
```

A client executes a query and polls for result completion. The server returns a FlightInfo representing the state of the query execution up to that point. 

```mermaid
sequenceDiagram
    Client->>Server: PollFlightInfo(FlightDescriptor)
    Server->>Client: PollInfo(FlightDescriptor', FlightInfo)
    Client->>Server: PollFlightInfo(FlightDescriptor')
    Server->>Client: PollInfo(FlightDescriptor'', FlightInfo)
    Client->>Server: PollFlightInfo(FlightDescriptor'')
    Server->>Client: PollInfo(_, FlightInfo)
```

The server:
* Must respond with the complete FlightInfo each time, not just the delta between the previous and current FlightInfo.
* Should respond as quickly as possible on the first call.
* Should not respond until the result would be different from last time. (That way, the client can “long poll” for updates without constantly making requests. Clients can set a short timeout to avoid blocking calls if desired.) 
* May respond by only updating the `PollInfo.progress` value (though it shouldn’t spam the client with updates).
* Should recognize a `PollInfo.flight_descriptor` that is not necessarily the latest (in case the client misses an update in between).
* Should only append to the endpoints in `FlightInfo` each time. (Otherwise the client has to do extra work to identify what endpoints it has and hasn’t seen.) 

  When `FlightInfo.ordered` is set, this means the server returns endpoints in order.
* Should return an error status instead of a response if the query fails. The client should not retry the request (except for TIMED_OUT and UNAVAILABLE, which may not originate from the server). 

### Prior Art

* [Amazon Redshift](https://docs.aws.amazon.com/redshift/latest/mgmt/data-api-calling.html): executing a query gives an ID that can be used to check the query status and fetch results.
* [Google BigQuery Storage](https://github.com/googleapis/googleapis/blob/master/google/cloud/bigquery/storage/v1/storage.proto): you explicitly create a “read session”, after which you can read subsets of the response with further requests. There is no “query execution time” since BigQuery Storage only queries tables. Instead, running a query (with the base BigQuery API) will cache the result in a table that can be read via BigQuery Storage.
* [Snowflake](https://docs.snowflake.com/en/developer-guide/sql-api/handling-responses.html#understanding-the-flow-of-execution): short queries return synchronously. Longer queries require polling for completion of the query. You cannot retrieve any results until the query is complete.

### Component(s)

C++, FlightRPC, Format, Go, Java

### PR DESCRIPTION
**Pull request by kou**
### Rationale for this change

In Flight RPC, FlightInfo includes addresses of workers alongside result partition info. This lets clients fetch data directly from workers, in parallel or even distributed across multiple machines. But this also comes with tradeoffs.

Queries generally don't complete instantly (as much as we would like them to). So where can we put the 'query evaluation time'?

* In `GetFlightInfo`: block and wait for the query to complete.
  * Con: this is a long-running blocking call, which may fail or time out. Then when the client retries, the server has to redo all the work.
  * Con: parts of the result may be ready before others, but the client can't do anything until everything is ready.
* In `DoGet`: return a fixed number of partitions
  * Con: this makes handling worker failures hard. Systems like Trino support fault-tolerant execution by replacing workers at runtime. But GetFlightInfo has already passed, so we can't notify the client of new workers.
  * Con: we have to know or fix the partitioning up front.

Neither solution is optimal.

### What changes are included in this PR?

We can address this by adding a retryable version of `GetFlightInfo`: `PollFlightInfo(FlightDescriptor)`

`PollFlightInfo` returns `PollInfo`:

```proto
message PollInfo {
  // The currently available results so far.
  FlightInfo info = 1;
  // The descriptor the client should use on the next try.
  // If unset, the query is complete.
  FlightDescriptor flight_descriptor = 2;
  // Query progress. Must be in [0.0, 1.0] but need not be
  // monotonic or nondecreasing. If unknown, do not set.
  optional double progress = 3;
  // Expiration time for this request. After this passes, the server
  // might not accept the retry descriptor anymore (and the query may
  // be cancelled). This may be updated on a call to PollFlightInfo.
  google.protobuf.Timestamp expiration_time = 4;
}
```

See the documentation changes for details of them:
http://crossbow.voltrondata.com/pr_docs/36946/format/Flight.html#downloading-data-by-running-a-heavy-query

### Are these changes tested?

Yes.

This has C++, Go and Java implementations and an integration test with them.

### Are there any user-facing changes?

Yes.
* Closes: #36155

### CODE REVIEW

File: docs/source/format/Flight.rst [hunk 1]
```diff
@@ -153,6 +153,70 @@ A client that wishes to download the data would:
 
 .. _google.protobuf.Timestamp: https://protobuf.dev/reference/protobuf/google.protobuf/#timestamp
 
+Downloading Data by Running a Heavy Query
+-----------------------------------------
+
+A client may need to request a heavy query to download
+data. ``GetFlightInfo`` isn't returned until the requested query is
+completed. It means that the client is blocked. The client can use
+``PollFlightInfo`` instead of ``GetFlightInfo``:
```
Comment (lidavidm) [inline docs/source/format/Flight.rst:162]:
```suggestion
A client may need to request a heavy query to download
data. However, ``GetFlightInfo`` doesn't return until the query
completes, so the client is blocked. In this situation, the client
can use ``PollFlightInfo`` instead of ``GetFlightInfo``:
```

File: docs/source/format/Flight.rst [hunk 2]
```diff
@@ -153,6 +153,70 @@ A client that wishes to download the data would:
...
+   A server should not respond until the result would be different
+   from last time. That way, the client can "long poll" for updates
+   without constantly making requests. Clients can set a short timeout
+   to avoid blocking calls if desired.
+
+   ``RetryInfo.progress`` may be set. It represents progress of the
+   query. If it's set, the value must be in ``[0.0, 1.0]``. The value
+   must not be monotonic or nondecreasing. A server may respond by
+   only updating the ``RetryInfo.progress`` value though it shouldn't
+   spam the client with updates.
```
Comment (lidavidm) [inline docs/source/format/Flight.rst:203]:
```suggestion
   ``RetryInfo.progress`` may be set. It represents progress of the
   query. If it's set, the value must be in ``[0.0, 1.0]``. The value
   is not necessarily monotonic or nondecreasing. A server may respond by
   only updating the ``RetryInfo.progress`` value, though it shouldn't
   spam the client with updates.
```

**Review by lidavidm — APPROVED** [review]
LGTM, thank you! I mostly left comments on wording. I can follow up with Java/Python later.

**Comment by lidavidm** [comment]
Java: https://github.com/kou/arrow/pull/14

File: go/arrow/internal/flight_integration/scenario.go [hunk 3]
```diff
@@ -1131,6 +1134,112 @@ func (tester *expirationTimeRenewFlightEndpointScenarioTester) RunClient(addr st
...
+	}
+	if info.Progress == nil {
+		return fmt.Errorf("progress is missing: %s", info.String())
+	}
+	if !(0.0 <= *info.Progress && *info.Progress <= 1.0) {
+		return fmt.Errorf("invalid progress: %s", info.String())
+	}
+	if info.ExpirationTime == nil {
+		return fmt.Errorf("expiration time is missing: %s", info.String())
+	}
```
Comment (zeroshade) [inline go/arrow/internal/flight_integration/scenario.go:1218]: The typical Go style would be to do this as a switch:

```go
switch {
case info.FlightDescriptor == nil:
        return fmt.Errorf.....
case info.Progress == nil:
        return fmt.Errorf.....
case !(0.0 <= *info.Progress && *info.Progress <= 1.0):
        return fmt.Errorf.....
case info.ExpirationTime == nil:
        return fmt.Errorf.....
}
```

File: go/arrow/internal/flight_integration/scenario.go [hunk 4]
```diff
@@ -1131,6 +1134,112 @@ func (tester *expirationTimeRenewFlightEndpointScenarioTester) RunClient(addr st
...
+			info.String())
+	}
+	if math.Abs(*info.Progress-1.0) > 1e-5 {
+		return fmt.Errorf("progress for finished query isn't 1.0: %s",
+			info.String())
+	}
+	if info.ExpirationTime != nil {
+		return fmt.Errorf("expiration time must not be set for finished query: %s",
+			info.String())
+	}
```
Comment (zeroshade) [inline go/arrow/internal/flight_integration/scenario.go:1238]: same as above

**Review by zeroshade — COMMENTED** [review]
LGTM just a style comment

Thanks much for doing the Go side of this!

**Comment by kou** [comment]
Thanks for suggesting wordings! They are very helpful for me.
I've applied all of them!

I'll update the Go style later.

**Comment by kou** [comment]
@github-actions crossbow submit preview-docs

File: go/arrow/internal/flight_integration/scenario.go (hunk 3 above)
Comment (kou) [reply go/arrow/internal/flight_integration/scenario.go]: Thanks!
I didn't know the syntax!

Done.

File: cpp/src/arrow/flight/middleware.h [hunk 5]
```diff
@@ -57,6 +57,7 @@ enum class FlightMethod : char {
   DoAction = 7,
   ListActions = 8,
   DoExchange = 9,
+  PollFlightInfo = 10,
```
[the remark this answers is not in the document]
Comment (kou) [reply cpp/src/arrow/flight/middleware.h:60]: We'll do this in another PR. Issue for this: https://github.com/apache/arrow/issues/36954

**Comment by kou** [comment]
http://crossbow.voltrondata.com/pr_docs/36946/format/Flight.html#downloading-data-by-running-a-heavy-query

**Review by zeroshade — APPROVED** [review]

File: format/Flight.proto [hunk 6]
```diff
@@ -343,6 +368,43 @@ message FlightInfo {
...
+  FlightInfo info = 1;
+
+  /*
+   * The descriptor the client should use on the next try.
+   * If unset, the query is complete.
+   */
+  FlightDescriptor flight_descriptor = 2;
+
+  /*
+   * Query progress. Must be in [0.0, 1.0] but need not be
```
Comment (alamb) [inline format/Flight.proto:395]:
```suggestion
   * Query progress. If known, must be in [0.0, 1.0] but need not be
```

File: format/Flight.proto [hunk 7]
```diff
@@ -65,6 +65,32 @@ service FlightService {
    */
   rpc GetFlightInfo(FlightDescriptor) returns (FlightInfo) {}
 
+  /*
+   * For a given FlightDescriptor, start a query and get information
+   * to poll the updated statuses. This is a useful interface if the
```
Comment (alamb) [inline format/Flight.proto:70]:
```suggestion
   * to poll its execution status. This is a useful interface if the
```

File: format/Flight.proto [hunk 8]
```diff
@@ -65,6 +65,32 @@ service FlightService {
...
+   * query may be a long-running query. The first PollFlightInfo call
+   * should return as quickly as possible. (GetFlightInfo doesn't
+   * return until the query is complete.)
+   *
+   * A client can consume the current results so far before
+   * the query is completed. See RetryInfo.info for details.
```
Comment (alamb) [inline format/Flight.proto:76]:
```suggestion
   * A client can consume any available results before
   * the query is completed. See RetryInfo.info for details.
```

**Review by alamb — APPROVED** [review]
TLDR is I read this proposal and it makes sense to me. Thank you for working on this. I only reviewed the proto and text, not the implementations

I left some minor suggested wording changes that I believe increase the precision and clarity of language, but I don't think they are necessary

Nice work

**Comment by alamb** [comment]
prior to merging this PR I think we need to do a formal vote on the change to the spec on the mailing list

File: format/Flight.proto [hunk 9]
```diff
@@ -343,6 +368,43 @@ message FlightInfo {
   bool ordered = 6;
 }
 
+/*
+ * The information to process a long-running query.
+ */
+message RetryInfo {
```
[the remark this answers is not in the document]
Comment (kou) [reply format/Flight.proto:374]: Thanks. It's a good point.
`RetryInfo` means info to retry polling but I can understand what you thought.

@lidavidm What do you think about this? My suggestion is `PollInfo` to align `PollFlightInfo()`.

**Review by kou — COMMENTED** [review]
> I left some minor suggested wording changes that I believe increase the precision and clarity of language, but I don't think they are necessary

Thanks! It's very helpful! I've merged your suggestions.

> prior to merging this PR I think we need to do a formal vote on the change to the spec on the mailing list

Yes. You're right. I'll do it in this week or next:

https://lists.apache.org/thread/qcjpcw6m3p15wqxp6n6rqzlx01v1fl3v

> Next:
>
> I'll start a vote for this proposal after we reach a consensus
> on this proposal.

File: format/Flight.proto (hunk 9 above)
[the remark this answers is not in the document]
Comment (lidavidm) [reply format/Flight.proto:374]: PollInfo sounds good.

[the remark this answers is not in the document]
Comment (kou) [reply format/Flight.proto]: OK. I've renamed to `PollInfo`.

**Comment by kou** [comment]
The discussion thread: https://lists.apache.org/thread/qcjpcw6m3p15wqxp6n6rqzlx01v1fl3v
The vote thread: https://lists.apache.org/thread/rwwqws6ghymx07qtz8d1jy5hsgljq96y
The vote carried: https://lists.apache.org/thread/9ty2vp3xnfnzcpl3wh5bnzbfszogomx2
